lib/Dialect/OM/Transforms/LinkModules.cpp

```diff
@@ -0,0 +1,224 @@
+  for (auto module : toplevelModule.getOps<ModuleOp>()) {
+    auto name = module->getAttrOfType<StringAttr>("om.namespace");
+    // Use `counter` if the namespace is not specified beforehand.
+    if (!name) {
+      name = StringAttr::get(module.getContext(), "module_" + Twine(counter++));
+      module->setAttr("om.namespace", name);
+    }
+    modules.emplace_back(module);
+  }
+
+  if (failed(failableParallelForEach(&getContext(), modules, [](auto &info) {
+        // Collect local information.
+        return info.initialize();
+      })))
+    return signalPassFailure();
+
+  // 2. Symbol resolution. Check that there is exactly single definition for
+  //    public symbols and rename private symbols if necessary.
+
+  // Global namespace to get unique names to symbols.
+  Namespace nameSpace;
+  // A map from a pair of enclosing module op and old symbol to a new symbol.
+  SymMappingTy symMapping;
+
+  // Construct a global map from symbols to class operations.
+  llvm::MapVector<StringAttr, SmallVector<ClassLike>> symbolToClasses;
+  for (const auto &info : modules)
+    for (auto &[name, op] : info.symbolToClasses) {
+      symbolToClasses[name].push_back(op);
+      // Add names to avoid collision.
+      (void)nameSpace.newName(name.getValue());
+    }
+
+  // Resolve symbols. We consider a symbol used as an external module to be
+  // "public" thus we cannot rename such symbols when there is collision. We
+  // require a public symbol to have exactly one definition so otherwise raise
+  // an error.
+  for (auto &[name, classes] : symbolToClasses) {
+    bool existExternalClass = false;
+    size_t countDefinition = 0;
+    for (auto op : classes) {
+      if (isa<ClassExternOp>(op))
+        existExternalClass = true;
+      else
+        ++countDefinition;
+    }
+
+    // There must be exactly one definition if the symbol was referred by an
+    // external class.
+    if (existExternalClass && countDefinition != 1) {
+      SmallVector<Location> classExternLocs;
+      SmallVector<Location> classLocs;
+      for (auto op : classes)
+        (isa<ClassExternOp>(op) ? classExternLocs : classLocs)
+            .push_back(op.getLoc());
+
+      auto diag = emitError(classExternLocs.front())
+                  << "OM linker error: class " << name
+                  << " is declared as an external class but "
+                  << (countDefinition == 0 ? "there is no definition"
+                                           : "there are multiple definitions");
+      for (auto loc : ArrayRef(classExternLocs).drop_front())
+        diag.attachNote(loc) << "class " << name << " is declared here as well";
+
+      if (countDefinition != 0) {
+        // There are multiple definitions.
+        for (auto loc : classLocs)
+          diag.attachNote(loc) << "class " << name << " is defined here";
+      }
+      signalPassFailure();
+      return;
+    }
+
+    // TODO: Raise errors if linked external modules are not compatible with the
```

Comment:
I guess here it would make sense to have more strict checks? Besides the presence of expected fields, which the verifier enforces, we could be more strict at this point and check that the fields defined in the `class.extern` have the same names and types as the fields defined in each `class`?

Reply:
Yeah I think it's definitely right direction to check fields strictly. I'll check field mismatches here.

Reply:
Done by https://github.com/llvm/circt/pull/5743/commits/460bee0882ce332661a4f07b21ceb36c345b9faa.